Create records for appending tables

hyperrealist opened the pull request:
Fixes #18

- [x] records added
- [x] tests added

LuisFSegalla reviewed (commented):
It looks good for me. I'm not so sure how the newest versions of FastCS are implementing the attributes handling, but I can't see anything obviously wrong with the code. I'm happy to test it against real hardware whenever you're available to do so!

shihab-dls commented:
This projects going to be updated to use FastCS 0.11 once #17 is in; would you mind rebasing over that branch, @hyperrealist? I can help, if there are ambiguous merge conflicts.

coretl commented on docs/conf.py:
```diff
@@ -193,3 +193,11 @@
+
+nitpick_ignore = [
+    ("py:class", "pandablocks.responses.TableFieldInfo"),
+]
```
These are defined above in the file already, please can you merge with them

hyperrealist replied:
Some of these came from rebasing on #17. @shihab-dls

coretl commented on docs/conf.py:
```diff
@@ -193,3 +193,11 @@
+    ("py:class", "pandablocks.responses.TableFieldInfo"),
+]
+
+suppress_warnings = ["docutils"]
```
What was this needed for? Please add comment to explain

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -359,17 +367,81 @@ def _make_table_field(
+            # CLEAR
+            parent_block.add_attribute(
+                panda_name + PandaName(sub_field="CLEAR"),
+                AttrW(
```
This should be a `Command`

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -359,17 +367,81 @@ def _make_table_field(
+                        )
+                    ),
+                    description="Current table mode.",
+                    group=WidgetGroup.OUTPUTS.value,
```
Can you put this in `PARAMETERS` please so it appears under the table

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -359,17 +367,81 @@ def _make_table_field(
+                AttrR(
+                    Int(),
+                    description="Number of queued lines not yet processed.",
+                    group=WidgetGroup.READBACKS.value,
```
Can you put this in `PARAMETERS` please so it appears under the table

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -359,17 +367,81 @@ def _make_table_field(
+                    Bool(),
+                    description="Clear the table.",
+                    io_ref=DefaultFieldIORef(
+                        panda_name, self._raw_panda.put_value_to_panda
```
This should call `client.send(Put(panda_name, []))` which sets it to an empty table

@shihab-dls I don't know how to make an IORef do that

shihab-dls replied:
I think you'd want to do something like this instead:
```python
async def clear_table() -> None:
    await self._raw_panda.put_value_to_panda(panda_name, attribute.datatype, [])

parent_block.clear = Command(clear_table)
```
Although, it looks like the widget made for commands sends "False" to the `SharedPV`, which raises an error, which we should look at. Putting `True` to the command PV will clear the table.

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -359,17 +367,81 @@ def _make_table_field(
             Table(structured_datatype),
-            handler=TableFieldHandler(panda_name),
+            io_ref=TableFieldIORef(
+                panda_name, field_info, self._raw_panda.put_value_to_panda
```
This should check the value of the NEXT_WRITE attribute, and decide whether to `Put`, `Append` or `Append(last=True)` based on its value

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -359,17 +367,81 @@ def _make_table_field(
+                    Enum(enum.Enum("NextWrite", ["Replace", "Append", "Append Last"])),
+                    description="What the next write will do to the table.",
+                    io_ref=DefaultFieldIORef(
+                        panda_name, self._raw_panda.put_value_to_panda
```
This should not write to the PandA, it should just store the value

hyperrealist committed:
bump copier template

hyperrealist committed:
fix container build CI

hyperrealist committed:
CR: fix typos

hyperrealist committed:
CR: fix async bug

hyperrealist committed:
Add table NEXT_WRITE dispatch and append write support

hyperrealist committed:
Add tests for clear command and append wrapper coverage

coretl commented on src/fastcs_pandablocks/panda/io/table.py:
```diff
@@ -17,13 +18,25 @@
+class NextWrite(enum.Enum):
+    Replace = "Replace"
+    Append = "Append"
+    Append_Last = "Append Last"
```
Nit: these should be uppercase

```suggestion
    REPLACE = "Replace"
    APPEND = "Append"
    APPEND_LAST = "Append Last"
```

coretl commented on src/fastcs_pandablocks/panda/blocks/blocks.py:
```diff
@@ -385,15 +386,65 @@ def _make_table_field(
+                panda_name + PandaName(sub_field="MODE"),
+                AttrR(
+                    Enum(
+                        enum.Enum(
```
This is an integer valued enum. The values from PandA are strings.
@shihab-dls remind me what fastcs uses the names/values of an Enum for?

coretl commented on tests/test_blocks.py:
Noooo, look at all the mocks! Please tell it to rewrite this in the style of the other tests, only mock out the client connection, everything else is real code.

hyperrealist committed:
Refine table enums and rewrite block tests with real RawPanda

hyperrealist committed:
Fix CLEAR command naming for multiple has_mode tables

- Changed from hard-coded parent_block.clear to per-table CLEAR commands
- Each table field now gets its own CLEAR command using qualified PandaName
- Uses setattr() with attribute_name derived from table field + CLEAR sub_field
- Prevents overwrites when multiple has_mode tables exist on same block
- Clients can now address clear action per table via expected name (issue #18)
- Added test_make_table_field_multiple_has_mode_tables to verify independence
